### ISSUE TITLE
sitewide 508-defect-3, design system, forms, components [SCREENREADER, FOCUS MANAGEMENT] While on "Please confirm your address" screen, the content from the previous "Edit mailing address" is read by screen reader

### ISSUE THREAD

**Issue by jenstrickland** [department-of-veterans-affairs/va.gov-team#5865]
```diff
! Team affected: vsa-authenticated-xp; Project: Address validation
```

## Description

After submitting the address edit, the "Edit mailing address" form reads again, although the screen displayed is "Please confirm your address".

It may be necessary to have a meeting where I show team this behavior on the Windows machine with IE11 and JAWS. 

> Note: this issue also happens when a user logs in via ID.me, where there is a button similar to the progress button. After logging in, the screen reader continues to read the previous login pages, although the page shown in the browser is the landing page.

## Point of Contact

**VFS Point of Contact:** Jennifer

## Acceptance Criteria

As a screen reader user, I want to hear the content of the screen I am currently on.

## Environment

* Operating System: Windows
* Browser: IE11
* Screenreading device: JAWS
* Server destination: staging

## Steps to Recreate

<!--
1. Enter `https://staging.va.gov/profile` in browser
2. Start screenreading device listed in Environment
3. Navigate to edit the Mailing address
4. Make a change to the address, and select Update
5. Verify that the screen displays "Please confirm your address" while hearing "Edit mailing address" content

**Comment by Samara-Strauss** [comment]
Going to triage this over to @bradl3yC. @bradl3yC — can you meet with @jenstrickland to get more information on this issue?

**Comment by Samara-Strauss** [comment]
Also, I'm going to add this to our sprint objectives since fixing all QA/508 bugs is our top objective.

**Comment by bradl3yC** [comment]
@1Copenut have you ever seen this before?  Its basically re-reading a screen thats no longer on the DOM.  It only happens in IE11.  Steps to reproduce: Edit address in user profile - click submit - address validation modal opens - JAWS re-reads edit address modal

**Comment by Samara-Strauss** [comment]
@bradl3yC can you add an estimate to this when you have a chance?

**Comment by 1Copenut** [comment]
@jenstrickland && @bradl3yC 
I took at look at this one after Brad reached out to me on Slack. I think I have an answer for you as to why you're hearing a lot of superfluous and sometimes wrong information read out.

Each of the content blocks has an `aria-live="polite"` attribute. This means each block is being flagged as needing read out when something changes. Screen readers (VO in my case) read everything on first load because those containers are empty on first pass, then hydrated with API calls.

When editing the mailing address, the content block is stripped of all content, with an update message added, then removed. This triggers the screen reader to re-read the entire block.

A better alternative would be to remove all of the aria-live declarations, and include a single message that we're updating the user's information. We use a similar treatment on multi-page forms (it's below the Continue button) and could be co-opted here. I'll defer to @Samara-Strauss on the right way to tackle the problem visually, but the aural oddity is almost certainly due to the live regions.

---
![Screen Shot 2020-02-20 at 3.39.46 PM.png](https://images.zenhubusercontent.com/5ac217b74b5806bc2bcd3fc8/c260f232-ebd6-4c3b-b047-91828f64d0fc)

**Comment by Samara-Strauss** [comment]
Does this require a visual update, or could the code fix go out before a design update? Sorry if that's the stupid question, it's just not clear to me how this would affect the UI based on the above. Might just need some additional clarification.

**Comment by 1Copenut** [comment]
@Samara-Strauss I'm not sure tbh. The fix for screen readers is to remove all of the `aria-live` regions so the devices will stop reading out all of the content blocks every time there's a change. That fix is overly harsh though. Screen readers do need to know when the new information is saved, but right now we have a blue message appearing inside each content block, then disappearing. So we can't just focus on the block or the update message. I screen capped a few views to create a short narrative.

I think a long-term solution is to put the update message in one place on the page, and have it fire on information save/update/remove, and announce to screen readers then as well. But, that requires a design change, so I'm erring on the side of just highlighting the potential trouble spots.

---
![sr-live-regions](https://user-images.githubusercontent.com/934879/74984905-13f66400-53fc-11ea-941e-89c87c40d40a.jpg)

**Comment by Samara-Strauss** [comment]
> I think a long-term solution is to put the update message in one place on the page, and have it fire on information save/update/remove, and announce to screen readers then as well. But, that requires a design change, so I'm erring on the side of just highlighting the potential trouble spots.

We can fix this with the Profile 2.0 work that is in progress right now. Pulling in @tressaellen who is doing that work. **Tressa** let's make sure that the save/error/update state going forward is clearer and appears consistently in the same spot. I suggest reviewing the approach with @jenstrickland to make sure it doesn't cause the same issues as the save interaction now.

As for a more short term code fix, @1Copenut @bradl3yC let's move forward with the fix outlined above if it can happen separately from UI fixes, and assume we'll fix the UI with the upcoming profile 2.0 redesign. But please let us know if there are immediate UI changes that happen so we can either find a simple UI fix or punt the code updates down the line to the profile 2.0 build.

**Comment by 1Copenut** [comment]
@Samara-Strauss && @jenstrickland I agree with the long-term strategy.

In the short-term I propose removing the `aria-live` regions from these content blocks. We also need to return focus to the `Edit` button that opened the modal in the first place. This is not ideal but at least it gives assistive device users a consistent experience. By returning focus to the button that opened the modal initially, we give users enough context and orientation to read the next few lines and confirm their address, phone number, etc. was updated successfully.

Does that work for everyone?

**Comment by Samara-Strauss** [comment]
Works for me if it works for everyone else :)

**Comment by bradl3yC** [comment]
@1Copenut did the focus not return to edit?  I think its because its a modal that opens a modal.  I am not 100% sure how we're gonna solve that one.

**Comment by bradl3yC** [comment]
@1Copenut @jenstrickland @Samara-Strauss the findings regarding the `Edit` button not retaining focus may need to be moved to another ticket as it related to `formation-react`.  The code changes will need to be made in a completely separate rep.  Also, the `Edit` button does receive focus on all button clicks EXCEPT Update.

**Comment by jenstrickland** [comment]
@bradl3yC Regarding @1Copenut 's comment, "We also need to return focus to the Edit button that opened the modal in the first place.": 

This refers to when all the modals are closed and the user is returned to the Profile page with focus on the Edit link that launched the modal. Yes, this would be an entirely different ticket. I'll open that now, and edit this to include a link.

**Comment by Samara-Strauss** [comment]
Are we closing this ticket in favor of #6156 or do we still need to keep this ticket open?

**Comment by bradl3yC** [comment]
@Samara-Strauss we need this one too :)

**Comment by Samara-Strauss** [comment]
Got it @bradl3yC.

**Comment by jenstrickland** [comment]
Work related to this happening: https://github.com/department-of-veterans-affairs/vets-website/pull/11743

**Comment by jenstrickland** [comment]
The content is still not reading properly. When I hit edit it and go back to Edit mailing address, instead of reading the modal, I hear 575 20th Ave, San Francisco… the address content on the Profile page.

For testing, please let me know when you have something in staging, and we can hop on a call on my Windows laptop to test. JAWS bugs can be tricky.

**Comment by jenstrickland** [comment]
@bradl3yC Checked this morning, and it works properly. 🎉

**Comment by Samara-Strauss** [comment]
Also #6472

**Comment by jenstrickland** [comment]
Update: 
Trevor and I reviewed and saw the same behavior. I am going to recommend changing this to a 508-defect-1 — in order to help the team launch. This will require that this must be fixed immediately after launch. Possibilities include:
- Implement edit in place 
- Have the address validation only in a modal
- Ensure focus management prevents the focus from being trapped behind the modal
- Seek another interim fix for a short-term solution. There is work happening in Profile 2.0 that will address this, but that's a couple of months out.
This is not to say that it wasn't a 508-defect-0 to begin with, as it certainly is. However, we want to help the team launch. We spent a week trying to fix it, and the closest we came was removing the layered modals, but introduced a undesired ux change.

**Comment by Samara-Strauss** [comment]
Thanks. We have removed this from our current sprint objectives, and we'll prioritize it for post-launch.

**Comment by erikphansen** [comment]
@jenstrickland and I just did a screenshare re: this issue and even with the Edit Phone modal there are times when JAWS is reading the text in the modal long after the modal has closed. It seems like as screen updates (or React renders) happen due to prop changes, JAWS will collect a queue of screen updates it needs to reads and it insists on reading them even after they are no longer on the page. For example, if the  edit phone modal shows an invalid phone number form error and you quickly fix the error, JAWS will insist on reading that error and all other changes in the modal before it starts to read what is _actually_ on the screen.

Also a similar issue happens with the Find VA Benefits feature. If you go to `my-va/find-benefits`, select some benefits, and hit the Save Preferences button at the bottom of the page, you'll be redirected back to `my-va/` but JAWS will continue reading _everything_ that's on the previous page. It seems like clicking the `Save Preferences` button might trigger a re-read of the `my-va/find-benefits` page. I mention that because we use a `LoadingButton` for the Update button on the edit modals in the profile as well as the Save Preferences button on `my-va/find-benefits`. Perhaps that's a clue?

I think that fixing the Find VA Benefits issue I described above is a smaller issue that might be more easily solved and we could learn something there that could be applicable to the issues with the profile modals.

In summary:
- this issue does not seem to be _because of_ the new address validation work. we saw issues in both production where the feature doesn't exist as well as in staging with an Edit Phone modal.
- we might want to focus on fixing the Find VA Benefits JAWS issue first and see if that fix could apply to the edit modals.

**Comment by Samara-Strauss** [comment]
What I'm gathering from this is that this issue has to do with how JAWS works, and that it does not play well with modals.

- @jenstrickland @1Copenut I would love to see broader guidance about how to handle design with JAWS in mind throughout the site. This issue has been open for a month and commanded incredible resources from our team, and we don't have the bandwidth to put a ton of resources on this now or ever. I would imagine other teams feel the same, especially if they only have one FE dev (we're lucky to have two!). I know we are getting pre-emptive practices in place with the VSP review cycle, and that reviews much earlier in the process might have caught something like this, but I still think some general JAWS guidance would be useful since it sounds like that's the main screenreader on the market for folks using PCs.
-  @erikphansen if you want to draft up a ticket(s) for the Find VA Benefits work, feel free to do so. TBH this is not a huge priority for our team and I don't want to sink a ton of resources and time into this especially since we'll need you helping to build out Profile 2.0, but we can prioritize it against the rest of our backlog if the work is documented there.

**Comment by erikphansen** [comment]
I think that if i was able to test with JAWS locally myself it would go a long way to speeding this all along. And I suspect that the issues we're seeing with JAWS reading modals that are no longer on the screen and JAWS reading content on a page that's no longer being shown are related. Once we can figure out how to make one happy I think we'll learn a lot and be able to resolve similar issues with JAWS reading content that's been removed from the screen.

**Comment by sshein** [comment]
Hi Erik - I just spoke with Andre about potentially getting more of our developers the ability to test with JAWS locally. What would you need to make that happen?

**Comment by erikphansen** [comment]
We lack legal licenses to run it on our machines. I think that with a license I could then run it in a Windows VM on my Mac but I'm not sure about that. Step one is just getting another license so we can run it on more machines.

**Comment by erikphansen** [comment]
I mentioned to Jen and Trevor that we should talk to the maker of JAWS to see if they'd be amenable to giving us cheap or free licenses to help us make sure VA.gov plays well with their product.

**Comment by jenstrickland** [comment]
I spoke with Trevor this afternoon, asked how many JAWS-specific issues he's seen. He said he's seen three in his two years. 

Given the learning curve for using JAWS, and the rarity of JAWS-specific issues, although it is a nice-to-have, it may not be as necessary.

**Comment by 1Copenut** [comment]
A lot of the issues we see are not specific to JAWS, and can usually be re-created with IE11 + NVDA, or Firefox + NVDA. If the goal is to familiarize developers with screen readers, we'll get more mileage out of NVDA and IE11 on a VM for first testing. Engineers can always reach out to Jennifer or myself to do follow up tests with IE11 + JAWS if they're seeing something unique. I'd venture a number of these issues are more related to IE11 than JAWS, but that's a non-data-driven hunch.

**Comment by jenstrickland** [comment]
Trevor updated the 508 Defect Severity Rubric, so I just modified the defect rating on this issue to 508-defect-2, to allow launch without resolving this. Since this is likely a more complex issue with the loading spinner button (based on troubleshooting with Erik), it **may** require rearchitecting.  With the Profile 2.0 redesign, we should validate with JAWS+IE11 as soon as possible.

**Comment by bradl3yC** [comment]
This shouldn't be related to the spinner.  As part of my original troubleshooting, I removed the prop change that triggered the spinner and the issue still existed.

**Comment by erikphansen** [comment]
But this issue with content being read after it's removed from the DOM (with the profile modals and the Find VA Benefits options) are only with IE11 + JAWS, correct? Or can it also be replicated with free software?

**Comment by bradl3yC** [comment]
@erikphansen this seems mostly isolated to IE11 / JAWS

**Comment by jenstrickland** [comment]
@erikphansen I tested with NVDA+IE11 as well as JAWS+Chrome, and there was no issue, so this particular issue seems isolated to JAWS+IE11.

**Comment by jenstrickland** [comment]
@erikphansen and I were able to replicate the behavior on the My VA Find Benefits screen, with the spinner button. Trevor had a suspicion "of the React `modal` component, Redux sending too large an object back, React rendering the entire modal, not the button, and JAWS interpreting the update as a whole new modal." 

We are concerned we could see no real progress on this issue for weeks, which is why I downgraded the defect level, hoping to provide enough time to architect a new experience in the next few sprints. I acknowledge this may take longer than we'd like, as this is more time-consuming to troubleshoot due to the remote collaboration. I am prioritizing checking fixes over everything else — the moment I get a notice there is something to check in staging, I drop whatever I'm doing to test.

**Comment by bradl3yC** [comment]
The developers really need access to JAWs to expedite this process.  The guess and check method of fixing this is not sustainable

**Comment by erikphansen** [comment]
@jenstrickland can you confirm that the issue we had with Find VA Benefits, with the content of the checkbox screen being read after returning to the main Find VA Benefits page, is also with JAWS only? that could go a long way in helping us get to the bottom of what's really tripping up JAWS in IE11.

**Comment by jenstrickland** [comment]
@erikphansen Yes, when I checked earlier, I meant that I checked both the Profile modals and the My VA Find Benefits page. The issue only happens in JAWS+IE11, not with Chrome.

**Comment by bradl3yC** [comment]
Is this not a JAWs issue then? If other screen readers read the content just fine on IE11?

**Comment by jenstrickland** [comment]
For clarity, the issue happens only with the **JAWS+IE11 combination**, and it happens on both the Profile modals and the Find Benefits page.

It does **not** happen with:
- JAWS+Chrome
- JAWS+Firefox
- NVDA+IE11

**Comment by 1Copenut** [comment]
A short-term strategy might be to have engineers download a Windows VM from modern.ie and run NVDA + Firefox in VirtualBox or some other hypervisor. JAWS-specific issues are rare, but they do happen. The longer term answer is to get the accessibility experts access to feature branches on Windows machines, so we're not waiting for code to be pushed up to Staging to test.

**Comment by Samara-Strauss** [comment]
Hi All – We're at a point where we need to discuss a path forward for this ticket. As I understand it, we have two issues here:

- The JAWS modal issue that was discovered during the address validation 508 review. From what I understand, this issue is actually not specific to address validation and could happen with any modal we have on our site.
- The JAWS issue around the "Find VA Benefits" tool that @erikphansen has offered to look into.

At this point, it sounds like the best thing is to have @erikphansen look into the "Find VA Benefits" tool issue and see what he can find. @mshea0606 is going to work with @erikphansen to create a ticket for that. 

However, I do not feel comfortable committing resources to trying to figure out the modal issue. Perhaps @erikphansen will discover something in the "Find VA Benefits" work that will be helpful in resolving the modal issue that was discovered in the profile, and if he does, great! But if not, it does not seem like our team can do this work effectively without a JAWS license of their own. Throwing things back and forth over the wall has proven inefficient, time-consuming, and frustrating. I can't ask @erikphansen and @bradl3yC to be spending their time this way when this process so far has not proven effective.

I want to be clear that I DO think this modal issue is important — our site needs to be just as usable for folks with screen readers as without. This is purely an issue of process, and needing to figure out a more effective path forward before I think it's wise for anyone — on our team or another person's team — to be trying to tackle these JAWS issues, especially one that is as nasty as the one documented here.

Please let me know if you have questions. Happy to discuss.

**Comment by jenstrickland** [comment]
@Samara-Strauss Since discovering this issue on the address validation work, as mentioned Erik and I discovered it on the Find VA Benefits tool, and I found it on the SSO on Friday. 

I completely agree that your team would be well-served to move forward with the Profile 2.0 effort, as this issue existed before and beyond your work. This bug needs to be addressed on a platform level, I believe. cc @1Copenut

**Comment by Samara-Strauss** [comment]
Thanks @jenstrickland sounds good to me.

When he's got time, @mshea0606 will work with @erikphansen to get the Find VA Benefits issue documented, and we'll work on that. Otherwise, I'll like you and @1Copenut manage this issue (5865) going forward.

**Comment by mshea0606** [comment]
@Samara-Strauss ticket #7324 has been created for the "Find VA Benefits" issue

**Comment by Samara-Strauss** [comment]
Thank you for letting me know @mshea0606. Can you please make sure it appears in the sprint objectives if we plan to tackle it this sprint (which I think we do)?

**Comment by caw310** [comment]
This needs to be retested on the minimum browsers but NOT IE11. 
Changing severity to Defect 3.

**Comment by SarahKay8** [comment]
Tested using Jaws+Chrome, Jaws+Firefox, Voiceover, NVDA+Chrome. I am seeing no issue on my end. Still believe this is an IE11 issue.

**Comment by caw310** [comment]
Closing since this is only an IE11 issue. There is a plan to stop supporting IE11.
